nengo/learning_rules.py

```diff
@@ -262,15 +308,49 @@ class Voja(LearningRuleType):
         Filter constant on activities of neurons in post population.
     """
 
-    error_type = 'scalar'
     modifies = 'encoders'
     probeable = ('post_filtered', 'scaled_encoders', 'delta')
 
     post_tau = NumberParam('post_tau', low=0, low_open=True, optional=True)
 
     def __init__(self, post_tau=0.005, learning_rate=1e-2):
         self.post_tau = post_tau
-        super(Voja, self).__init__(learning_rate)
+        super(Voja, self).__init__(learning_rate, size_in=1)
+
+
+class GenericRule(LearningRuleType):
+    """Learning rule implemented by generic Python function.
+
+    Parameters
+    ----------
+    function : callable
+        Accepts previous values of learned array and a data vector
+        as input, outputs a delta for the learned array.
+
+        Optionally the function may also accept a `params` argument, in which
+        case `model.params` will be passed to the function as well (can be used
+        to look up things like encoder values).
+    learning_rate : float
+        A scalar used to weight the output of the function
+    size_in : int
+        The dimensionality of the data vector input to the function
+    modifies : str
+        The signal targeted by the learning rule ('weights' or 'decoders')
+    pass_model_params : bool
+        Selects whether or not the compiled model parameters will be passed
+        to the learning rule function
+    """
+
+    probeable = ('target', 'data', 'delta')
+
+    def __init__(self, function, learning_rate=1.0, size_in=0,
+                 modifies="decoders", pass_model_params=False):
+        self.modifies = modifies
+        self.function = function
+        self.pass_model_params = pass_model_params
```

Comment:
Do you think the `modifies` parameter in the other learning rules should be a `Parameter` as well?  It's a bit odd, as in that in those other cases it is not something we expect to be set by the user while in this class it is.  But if only this one is a parameter, then if you print all the parameters of a learning rule (via e.g. `Config.all_defaults`) you'd get the impression that it's a unique attribute of `GenericRule` rather than a property of the `LearningRuleType` base class.